gnwmanager/gnw.py

```diff
@@ -504,6 +513,96 @@ def _flash_ext(
 
         self.wait_for_all_contexts_complete()
 
+    def _sd_write_file_chunk(
+        self,
+        path: str,
+        block: int,
+        total_blocks: int,
+        data: bytes,
+        blocking: bool = True,
+        compress: bool = True,
+    ) -> None:
+        """Low-level write data to fat filesystem.
+
+        Limited to RAM constraints (i.e. <256KB writes).
+
+        Parameters
+        ----------
+        path: str
+            Path of file to write
+        size: int
+            Number of bytes to write.
+        blocking: bool
+            Wait for action to be complete.
+        """
+        log.debug(f"gnw._sd_write_file_chunk: {path=} {len(data)=} {blocking=} {compress=}")
+
+        if not path:
+            return
+
+        if not data:
+            return
+        if len(data) > (256 << 10):
+            raise ValueError("Too large of data for a single write.")
+
+        if compress:
+            compressed_data = compress_lzma(data)
+            # If we are unable to compress meaningfully, don't bother.
+            if len(compressed_data) > (0.9 * len(data)):
+                compress = False
+        else:
+            compressed_data = b""
+
+        context = self.get_context()
+
+        log.debug("setting upload_in_progress.")
+        self.write_uint32("upload_in_progress", 1)
+
+        self.write_uint32(context["action"], actions["WRITE_FILE_TO_SD"])
+        self.write_str(context["dest_path"], path)
+        self.write_uint32(context["size"], len(data))
+        self.write_uint32(context["block"], block)
+        self.write_uint32(context["total_blocks"], total_blocks)
+
+        data_hash = sha256(data)
+        self.write_memory(context["expected_sha256"], data_hash)
+
+        if compress:
+            self.write_uint32(context["compressed_size"], len(compressed_data))
+            self.write_memory(context["buffer"], compressed_data)
+        else:
+            self.write_uint32(context["compressed_size"], 0)
+            self.write_memory(context["buffer"], data)
+
+        log.debug(f"Activating PROGRAM: {data_hash.hex()}")
+        self.write_uint32(context["ready"], self.context_counter)
+        self.context_counter += 1
+        log.debug(f"context_counter incremented to {self.context_counter}.")
+
+        log.debug("clearing upload_in_progress.")
+        self.write_uint32("upload_in_progress", 0)
+
+        if blocking:
+            self.wait_for_all_contexts_complete()
+
+    def sd_write_file(
+        self,
+        path: str,
```

Comment:
do we want some assertions on `path`? e.g. always start with `/`?

Reply:
probably ...

Reply:
resolved in commit b64787a6a800e7745ebda3f13b87d1c076cee2df